src/Storages/HDFS/AsynchronousReadBufferFromHDFS.cpp

```diff
@@ -0,0 +1,267 @@
+#include "AsynchronousReadBufferFromHDFS.h"
+
+#if USE_HDFS
+#include <mutex>
+#include <Common/logger_useful.h>
+#include <Storages/HDFS/HDFSCommon.h>
+#include <Storages/HDFS/ReadBufferFromHDFS.h>
+#include <Disks/IO/ThreadPoolRemoteFSReader.h>
+
+namespace CurrentMetrics
+{
+    extern const Metric AsynchronousReadWait;
+}
+
+namespace ProfileEvents
+{
+    extern const Event AsynchronousReadWaitMicroseconds;
+    extern const Event RemoteFSSeeks;
+    extern const Event RemoteFSPrefetches;
+    extern const Event RemoteFSCancelledPrefetches;
+    extern const Event RemoteFSUnusedPrefetches;
+    extern const Event RemoteFSPrefetchedReads;
+    extern const Event RemoteFSUnprefetchedReads;
+    extern const Event RemoteFSBuffers;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_SEEK_THROUGH_FILE;
+    extern const int SEEK_POSITION_OUT_OF_BOUND;
+    extern const int LOGICAL_ERROR;
+}
+
+AsynchronousReadBufferFromHDFS::AsynchronousReadBufferFromHDFS(
+    AsynchronousReaderPtr reader_, const ReadSettings & settings_, std::shared_ptr<ReadBufferFromHDFS> impl_)
+    : BufferWithOwnMemory<SeekableReadBuffer>(settings_.remote_fs_buffer_size)
+    , reader(reader_)
+    , priority(settings_.priority)
+    , impl(std::move(impl_))
+    , prefetch_buffer(settings_.remote_fs_buffer_size)
+    , read_until_position(impl->getFileSize())
+    , use_prefetch(settings_.remote_fs_prefetch)
+    , log(&Poco::Logger::get("AsynchronousReadBufferFromHDFS"))
+{
+    ProfileEvents::increment(ProfileEvents::RemoteFSBuffers);
+    if (use_prefetch)
+        prefetch();
+}
```

Comment:
let's call it separately, not in constructor

Reply:
Hive text format files are scanned sequentially from begin to end. So adding a prefetch before first `next` will help improve the performance. 

Where do you advice to move it to?